tests/linearizability/model.go

```diff
@@ -93,121 +96,77 @@ var etcdModel = porcupine.Model{
 }
 
 func step(state EtcdState, request EtcdRequest, response EtcdResponse) (bool, EtcdState) {
-	if request.Key == "" {
-		panic("invalid request")
-	}
-	if state.Key == "" {
-		return true, initState(request, response)
+	if len(state.PossibleValues) == 0 {
+		state.Key = request.Key
+		if ok, v := initValueRevision(request, response); ok {
+			state.PossibleValues = append(state.PossibleValues, v)
+		}
+		return true, state
 	}
 	if state.Key != request.Key {
-		panic("Multiple keys not supported")
+		panic("multiple keys not supported")
 	}
-	switch request.Op {
-	case Get:
-		return stepGet(state, request, response)
-	case Put:
-		return stepPut(state, request, response)
-	case Delete:
-		return stepDelete(state, request, response)
-	default:
-		panic("Unknown operation")
+	if response.Err != nil {
+		for _, v := range state.PossibleValues {
+			newV, _ := stepValue(v, request)
+			state.PossibleValues = append(state.PossibleValues, newV)
+		}
+	} else {
+		var i = 0
+		for _, v := range state.PossibleValues {
+			newV, expectedResponse := stepValue(v, request)
+			if expectedResponse == response {
+				state.PossibleValues[i] = newV
+				i++
+			}
+		}
+		state.PossibleValues = state.PossibleValues[:i]
```

Comment:
https://github.com/etcd-io/etcd/pull/14902/files#r1043092652 

Why do you still keep the entries [0, i)? I think we only need to keep the latest one.

Reply:
Because depending on request type, there might be multiple possible outputs. Imagine a situation:
* Init state: `[{"rev": 1, value:"1"}]`
* failed PUT("2") request. State: `[{"rev": 1, value:"1"}, {"rev": 2, value:"2"}]`
* TXN(if(value="3").then(value="4") that failed. State `[{"rev": 1, value:"1"}, {"rev": 2, value:"2"}]`
* Get. 

Both values `1` and `2` are equally probable. Txn doesn't cause state to collapse.

Reply:
> TXN(if(value="3").then(value="4") that failed

What we are talking about is the `else` branch (`response.Err == nil`), it means the request succeeds. So I assume "that failed" means the TXN's Compare condition is evaluated to `false`.  In this case, we might still need to cache the revision in the response header, because it's a real current state returned from the server side. 

The `stepValue` function doesn't support TXN yet, so I suggest not to enhance the implementation for a case we do not support yet. We can continue to enhance once we support TXN.  The TXN complicates the conversation and implementation. Can we keep it as simple as possible for now?

Reply:
No, there might be cases where multiple states are correct after transaction. For example:
* `get a -> 1, rev 1`. Possible states: [{"value": "1", "revision": 1}]
* `put a 2 -> error`. Possible states: [{"value": "1", "revision": 1}, {"value": "2", "revision": 2}]
* `put a 3 -> error`. Possible states: [{"value": "1", "revision": 1}, {"value": "2", "revision": 2}, {"value": "3", "revision": 2}, {"value": "3", "revision": 3}]
* `txn if(a=4).then(put a 5) -> false, rev 2`. Possible states: [{"value": "2", "revision": 2}, {"value": "3", "revision": 2}]

Because failed TXN doesn't know read the key, we don't know the end state. There are 4 possible histories here:
1. `get`, `txn`
2. `get`, `put 2`, `txn` 
3. `get`, `put 3`, `txn` 
4. `get`, `put 2`, `put 3`, `txn`

We can eliminate 1 and 4, because they would not match returned revision 2. But both 2 and 3 are possible as they both end with revision equal 2. So we don't know whether `put 2` or `put 3` was peristed.

Reply:
> * `txn if(=4).then(put 5) -> false, rev 2`. Possible states: [{"value": "2", "revision": 2}, {"value": "3", "revision": 2}]

This is a good point. We successfully removed one impossible entry from the `PossibleValues`.  Just as I mentioned in https://github.com/etcd-io/etcd/pull/14902#discussion_r1043849481, can we not consider TXN for now? Instead, let's consider supporting it after merging this PR?

Reply:
`state.PossibleValues = state.PossibleValues[:i]`

When there isn't any match, `then i == 0`, accordingly all values are cleaned up. This might not be correct.  As long as the request succeeds, we need to save the latest response (data and revision). WDYT?

Reply:
> What we are talking about is the else branch (response.Err == nil), it means the request succeeds. So I assume "that failed" means the TXN's

Yes, sorry for not clarifying it. 

> In this case, we might still need to cache the revision in the response header, because it's a real current state returned from the server side.

We don't need to cache. Major change from old code to the new code is that we get from partial to full traceability.
In previous model we just noted last failed request. If there was multiple failed requests in row, we stopped tracking the state until we get a successful request. Only think we validated was that revision increased. 

Proposed model does full tracing of requests. Which means for every failed request, we fork internal state to represent both possibilities of it being persisted and being lost. It has  exponential complexity, but guarantees correctness. 

We don't do any caching of revision as one of possible states should already have the revision. This is what filtering on `response == expectedResponse` does. It only leaves values that would end up the revision that we got from the etcd server.

 > The TXN complicates the conversation and implementation. Can we keep it as simple as possible for now?
 
 Don't agree, picking first match is not correct behavior. It's a difference of filtering vs picking random. Txn doesn't change that. If there are multiple states that would return correct response, we cannot just pick randomly from them. We need to allow both possibilities and assuming that on next `step` we will collapse the state.